File: LegacySource/settlement/room/main/RoomSingleton.cs
using System;
using System.Collections.Generic;
using System.IO;
using init.resources;
using init.sprite.UI;
using settlement.main;
using settlement.maintenance;
using settlement.room.main.construction;
using settlement.room.main.furnisher;
using settlement.room.main.util;
using settlement.room.sprite;
using snake2d;
using snake2d.util.datatypes;
using util.rendering;

namespace settlement.room.main
{
    public abstract class RoomSingleton : Room, Serializable
    {
        private static readonly long serialVersionUID = 1L;
        private int size;
        private readonly Rec tiles = new Rec(0, 0, 0, 0);
        private readonly Coo dataCoo = new Coo();
        private readonly Coo upperLeft = new Coo();
        private int dataTile;
        private int data;
        protected transient FurnisherItem item;

        protected static readonly transient RoomAreaWrapper wrap = new RoomAreaWrapper();
        protected static readonly transient RoomAreaWrapper wrapD = new RoomAreaWrapper();

        protected RoomSingleton(ROOMS m, RoomBlueprint p) : base(m, p, true) { }

        public final RoomSingleton Place(TmpArea area)
        {
            int mx = area.Mx();
            int my = area.My();

            area.ReplaceAndClear(this);
            IniHard(mx, my);
            ROOMA a = wrap.Init(this, mx, my);
            AddAction(a);

            SETT.ROOMS().Map.Init(a);
            IsolationSet(a.Mx(), a.My(), SETT.ROOMS().Isolation.GetProspect(Blueprint(), a, null));
            wrap.Done();
            SETT.MAINTENANCE().InitRoomDegrade(this, a.Mx(), a.My());

            return this;
        }

        protected override bool Render(Renderer r, ShadowBatch shadowBatch, RenderIterator i)
        {
            RoomSprite s = ROOMS().fData.sprite.Get(i.Tile());
            if (s != null)
                return s.Render(r, shadowBatch, ROOMS().fData.spriteData.Get(i.Tile()), i, GetDegrade(i.Tx(), i.Ty()), ROOMS().fData.candle.Is(i.Tile()));
            return false;
        }

        protected override bool RenderAbove(Renderer r, ShadowBatch shadowBatch, RenderIterator i)
        {
            RoomSprite s = ROOMS().fData.sprite.Get(i.Tile());
            if (s != null)
                s.RenderAbove(r, shadowBatch, ROOMS().fData.spriteData.Get(i.Tile()), i, GetDegrade(i.Tx(), i.Ty()));
            return false;
        }

        protected override bool RenderBelow(Renderer r, ShadowBatch shadowBatch, RenderIterator i)
        {
            RoomSprite s = ROOMS().fData.sprite.Get(i.Tile());
            if (s != null)
                s.RenderBelow(r, shadowBatch, ROOMS().fData.spriteData.Get(i.Tile()), i, GetDegrade(i.Tx(), i.Ty()));
            return false;
        }

        protected override AVAILABILITY GetAvailability(int tile)
        {
            FurnisherItemTile t = ROOMS().fData.tile.Get(tile);
            if (t != null)
                return t.availability;
            return AVAILABILITY.ROOM;
        }

        public override Furnisher Constructor()
        {
            return BlueprintI().Constructor();
        }

        public override bool DestroyTileCan(int tx, int ty)
        {
            return ROOMS().fData.availability.Get(tx, ty).player < 0 || ROOMS().fData.availability.Get(tx, ty).enemy < 0;
        }

        public override void DestroyTile(int tx, int ty)
        {
            IniHard(tx, ty);
            ConstructionInit init = new ConstructionInit(this, tx, ty, true);
            ROOMA ar = wrap.Init(this, tx, ty);

            foreach (COORDINATE c in ar.Body())
            {
                if (!ar.Is(c))
                    continue;
                PATH().availability.UpdateService(c.X(), c.Y());
                if (!TERRAIN().Get(c).Clearing().IsStructure())
                    TERRAIN().NADA.PlaceFixed(c.X(), c.Y());
            }
            wrap.Done();
            TmpArea a = Remove(tx, ty, false, this, true);
            ROOMS().construction.BreakIt(a, init, tx, ty);
        }

        public override int Area(int tx, int ty)
        {
            IniHard(tx, ty);
            return size;
        }

        public override TmpArea Remove(int tx, int ty, bool scatter, object obj, bool force)
        {
            IniHard(tx, ty);
            ROOMA a = wrap.Init(this, tx, ty);
            SETT.ROOMS().stats.broken().Remove(a.Mx(), a.My());
            RemoveAction(a);
            if (scatter)
                Deleter.ScatterMaterials(a, Constructor(), Upgrade(tx, ty), GetDegrade(tx, ty));

            foreach (COORDINATE c in a.Body())
            {
                if (!a.Is(c))
                    continue;
                SETT.LIGHTS().Remove(c.X(), c.Y());
                FLOOR().clearer.Clear(c.X(), c.Y());
                SETT.ROOMS().data.Set(a, c, 0);
                ConstructionData.dConstructed.Set(a, c, 1);
                ConstructionData.dFloored.Set(a, c, 1);
            }

            TmpArea ar = ROOMS().map.Delete(this, a.Mx(), a.My(), obj);

            wrap.Done();
            wrap.Clear();
            tiles.SetDim(0).MoveX1Y1(-1, -1);

            return ar;
        }

        protected void AddAction(ROOMA ins)
        {

        }

        protected void RemoveAction(ROOMA ins)
        {

        }

        public override ROOM_DEGRADER Degrader(int tx, int ty)
        {
            IniHard(tx, ty);
            degA = wrapD.Init(this, dataCoo.X(), dataCoo.Y());
            wrapD.Done();
            return degrader;
        }

        private transient ROOMA degA;
        private readonly Degrader degrader = new Degrader();

        private class Degrader : ROOM_DEGRADER, Serializable
        {
            private static readonly long serialVersionUID = 1L;

            public override int ResSize()
            {
                return item.group.blueprint.Resources();
            }

            public override int ResAmount(int i)
            {
                return (int)Math.Ceiling(item.cost2(i, Upgrade(degA.Mx(), degA.My())));
            }

            public override RESOURCE Res(int i)
            {
                return item.group.blueprint.Resource(i);
            }

            public override double DegRate()
            {
                return degradeResNeeded();
            }

            public override int GetData()
            {
                return data;
            }

            protected override void SetData(int v, bool b)
            {
                data = v;
            }

            public override Icon Icon()
            {
                return BlueprintI().iconBig();
            }

            private bool ini(int tx, int ty)
            {
                if (tiles.HoldsPoint(tx, ty) && item != null && item == ROOMS().fData.item.Get(tx, ty))
                    return true;

                ROOMS().fData.itemX1Y1(tx, ty, upperLeft, this);
                item = ROOMS().fData.item.Get(tx, ty);
                dataCoo.Set(upperLeft);
                dataCoo.Increment(item.firstX(), item.firstY());

                tiles.MoveX1Y1(upperLeft);
                tiles.SetWidth(item.width());
                tiles.SetHeight(item.height());
                size = item.width() * item.height();
                dataTile = dataCoo.X() + dataCoo.Y() * TWIDTH;
                data = ROOMS().data.Get(dataTile);

                return true;
            }

            protected void iniHard(int tx, int ty)
            {
                if (!ini(tx, ty))
                {
                    throw new RuntimeException(tx + " " + ty + " " + this + " " + SETT.ROOMS().map.Get(tx, ty));
                }
            }

            public override bool IsSame(int tx, int ty, int ox, int oy)
            {
                if (Blueprint().Is(tx, ty))
                {
                    iniHard(tx, ty);
                    return iss(ox, oy);
                }
                return false;
            }

            private static Coo upperLeftTest = new Coo();

            private bool iss(int tx, int ty)
            {
                if (tiles.HoldsPoint(tx, ty) && item == ROOMS().fData.item.Get(tx, ty))
                {
                    ROOMS().fData.itemX1Y1(tx, ty, upperLeftTest, this);
                    if (upperLeftTest.IsSameAs(upperLeft))
                        return true;
                }
                return false;
            }

            public override int Mx(int tx, int ty)
            {
                iniHard(tx, ty);
                return dataCoo.X();
            }

            public override int My(int tx, int ty)
            {
                iniHard(tx, ty);
                return dataCoo.Y();
            }

            public override int X1(int tx, int ty)
            {
                iniHard(tx, ty);
                return tiles.X1();
            }

            public override int Y1(int tx, int ty)
            {
                iniHard(tx, ty);
                return tiles.Y1();
            }

            public override int Width(int tx, int ty)
            {
                iniHard(tx, ty);
                return tiles.Width();
            }

            public override int Height(int tx, int ty)
            {
                iniHard(tx, ty);
                return tiles.Height();
            }
        }
    }
}

File: settlement/room/main/RoomSingleton.java
package settlement.room.main;

import static settlement.main.SETT.FLOOR;
import static settlement.main.SETT.PATH;
import static settlement.main.SETT.ROOMS;
import static settlement.main.SETT.TERRAIN;
import static settlement.main.SETT.TWIDTH;

import java.io.Serializable;

import init.resources.RESOURCE;
import init.sprite.UI.Icon;
import settlement.main.SETT;
import settlement.maintenance.ROOM_DEGRADER;
import settlement.path.AVAILABILITY;
import settlement.room.main.construction.ConstructionData;
import settlement.room.main.construction.ConstructionInit;
import settlement.room.main.furnisher.Furnisher;
import settlement.room.main.furnisher.FurnisherItem;
import settlement.room.main.furnisher.FurnisherItemTile;
import settlement.room.main.util.Deleter;
import settlement.room.main.util.RoomAreaWrapper;
import settlement.room.sprite.RoomSprite;
import snake2d.Renderer;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;
import snake2d.util.datatypes.Rec;
import util.rendering.RenderData.RenderIterator;
import util.rendering.ShadowBatch;

public abstract class RoomSingleton extends Room {

	private static final long serialVersionUID = 1L;
	private int size;
	private final Rec tiles = new Rec(0,0,0,0);
	private final Coo dataCoo = new Coo();
	private final Coo upperLeft = new Coo();
	private int dataTile;
	private int data;
	protected transient FurnisherItem item;
	
	protected static final transient RoomAreaWrapper wrap = new RoomAreaWrapper();
	protected static final transient RoomAreaWrapper wrapD = new RoomAreaWrapper();
	
	protected RoomSingleton(ROOMS m, RoomBlueprint p){
		super(m, p, true);
	}
	
	public final RoomSingleton place(TmpArea area) {
		
		int mx = area.mx();
		int my = area.my();
		
		area.replaceAndClear(this);
		iniHard(mx, my);
		ROOMA a = wrap.init(this, mx, my);
		addAction(a);
		
		
		SETT.ROOMS().map.init(a);
		isolationSet(a.mX(), a.mY(), SETT.ROOMS().isolation.getProspect(blueprint(), a, null));
		wrap.done();
		SETT.MAINTENANCE().initRoomDegrade(this, a.mX(), a.mY());
		
		
		return this;
	}
	
	@Override
	protected boolean render(Renderer r, ShadowBatch shadowBatch, RenderIterator i) {
		RoomSprite s = ROOMS().fData.sprite.get(i.tile());
		if (s != null)
			return s.render(r, shadowBatch, ROOMS().fData.spriteData.get(i.tile()), i, getDegrade(i.tx(), i.ty()), ROOMS().fData.candle.is(i.tile()));
		return false;
	}
	
	@Override
	protected boolean renderAbove(Renderer r, ShadowBatch shadowBatch, RenderIterator i) {
		RoomSprite s = ROOMS().fData.sprite.get(i.tile());
		if (s != null)
			s.renderAbove(r, shadowBatch, ROOMS().fData.spriteData.get(i.tile()), i, getDegrade(i.tx(), i.ty()));
		return false;
	}
	
	@Override
	protected boolean renderBelow(Renderer r, ShadowBatch shadowBatch, RenderIterator i) {
		RoomSprite s = ROOMS().fData.sprite.get(i.tile());
		if (s != null)
			s.renderBelow(r, shadowBatch, ROOMS().fData.spriteData.get(i.tile()), i, getDegrade(i.tx(), i.ty()));
		return false;
	}
	
	@Override
	protected AVAILABILITY getAvailability(int tile) {
		FurnisherItemTile t = ROOMS().fData.tile.get(tile);
		if (t != null)
			return t.availability;
		return AVAILABILITY.ROOM;
	}
	
	@Override
	public final Furnisher constructor() {
		return blueprintI().constructor();
	}
	
	@Override
	public boolean destroyTileCan(int tx, int ty) {
		return ROOMS().fData.availability.get(tx, ty).player < 0 || ROOMS().fData.availability.get(tx, ty).enemy < 0;
	}
	
	@Override
	public void destroyTile(int tx, int ty) {
		iniHard(tx, ty);
		ConstructionInit init = new ConstructionInit(this, tx, ty, true);
		ROOMA ar = wrap.init(this, tx, ty);
		
		
		for (COORDINATE c : ar.body()) {
			if (!ar.is(c))
				continue;
			PATH().availability.updateService(c.x(), c.y());
			if (!TERRAIN().get(c).clearing().isStructure())
				TERRAIN().NADA.placeFixed(c.x(), c.y());
		}
		wrap.done();
		TmpArea a = remove(tx, ty, false, this, true);
		ROOMS().construction.breakIt(a, init, tx, ty);
		
	}
	

	@Override
	public int area(int tx, int ty) {
		iniHard(tx, ty);
		return size;
	}
	
	@Override
	public final TmpArea remove(int tx, int ty, boolean scatter, Object obj, boolean force) {
		iniHard(tx, ty);
		ROOMA a = wrap.init(this, tx, ty);
		SETT.ROOMS().stats.broken().remove(a.mX(), a.mY());
		removeAction(a);
		if (scatter)
			Deleter.scatterMaterials(a, constructor(), upgrade(tx, ty), getDegrade(tx, ty));
		
		for (COORDINATE c : a.body()) {
			if (!a.is(c))
				continue;
			SETT.LIGHTS().remove(c.x(), c.y());
			FLOOR().clearer.clear(c.x(), c.y());
			SETT.ROOMS().data.set(a, c, 0);
			ConstructionData.dConstructed.set(a, c, 1);
			ConstructionData.dFloored.set(a, c, 1);
		}
		
		TmpArea ar = ROOMS().map.delete(this, a.mX(), a.mY(), obj);
		
		wrap.done();
		wrap.clear();
		tiles.setDim(0).moveX1Y1(-1, -1);
		
		return ar;
	}
	
	protected void addAction(ROOMA ins) {
		
	}
	
	
	protected void removeAction(ROOMA ins) {
		
	}
	
	@Override
	public ROOM_DEGRADER degrader(int tx, int ty) {
		iniHard(tx, ty);
		degA = wrapD.init(this,  dataCoo.x(),  dataCoo.y());
		wrapD.done();
		return degrader;
	}
	
	private transient ROOMA degA;
	private final Degrader degrader = new Degrader();
	
	private class Degrader extends ROOM_DEGRADER implements Serializable {
		
		/**
		 * 
		 */
		private static final long serialVersionUID = 1L;

		@Override
		public int resSize() {
			return item.group.blueprint.resources();
		}

		@Override
		public int resAmount(int i) {
			return (int) Math.ceil(item.cost2(i, upgrade(degA.mX(), degA.mY())));
		}

		@Override
		public RESOURCE res(int i) {
			return item.group.blueprint.resource(i);
		}

		@Override
		public double degRate() {
			return degradeResNeeded();
		}

		@Override
		public  int getData() {
			return data;
		}

		@Override
		protected void setData(int v, boolean realDegradeChange) {
			double old = get();
			data = v;
			
			ROOMS().data.set(degA, dataTile, data);
			if (old != get())
				degradeChange(dataCoo.x(), dataCoo.y(), old, get(), realDegradeChange);
		}

		@Override
		public int roomArea() {
			return item.width()*item.height();
		}
		
		@Override
		public double base() {
			return blueprintI().degradeRate();
		}

		@Override
		public double expenseRate() {
			return 1;
		}

		@Override
		public double rate(double bonus) {
			double am = 0;
			for (int ri = 0; ri < resSize(); ri++) {
				am += resAmount(ri);
			}
			return Degrader.rate(bonus, base(), isolation(dataCoo.x(), dataCoo.y()), am, area(dataCoo.x(), dataCoo.y()));
		}
	};
	
	@Override
	public int resAmount(int ri, int upgrade) {
		return (int) Math.ceil(item.cost2(ri, upgrade));
	}
	
	protected void degradeChange(int mx, int my, double oldD, double newD, boolean realDegradeChange) {
		
	}
	
	protected double degradeResNeeded() {
		return 0.5;
	}
	
	public abstract RoomBlueprintImp blueprintI();
	
	@Override
	public CharSequence name(int tx, int ty) {
		iniHard(tx, ty);
		if (blueprintI().constructor() != null && blueprintI().constructor().groups().size() > 1)
			return item.group.name;
		return blueprintI().info.name;
	}
	
	@Override
	public Icon icon() {
		return blueprintI().iconBig();
	}
	
	private final boolean ini(int tx, int ty) {
		if (tiles.holdsPoint(tx, ty) && item != null && item == ROOMS().fData.item.get(tx, ty))
			return true;
		
		ROOMS().fData.itemX1Y1(tx, ty, upperLeft, this);
		item = ROOMS().fData.item.get(tx, ty);
		dataCoo.set(upperLeft);
		dataCoo.increment(item.firstX(), item.firstY());
		
		tiles.moveX1Y1(upperLeft);
		tiles.setWidth(item.width());
		tiles.setHeight(item.height());
		size = item.width()*item.height();
		dataTile = dataCoo.x()+dataCoo.y()*TWIDTH;
		data = ROOMS().data.get(dataTile);
		
		return true;
	}
	
	protected void iniHard(int tx, int ty) {
		if (!ini(tx, ty)) {
			throw new RuntimeException(tx + " " + ty + " " + this + " " + SETT.ROOMS().map.get(tx, ty));
		}
	}
	
	@Override
	public boolean isSame(int tx, int ty, int ox, int oy) {
		if (blueprint().is(tx, ty)) {
			iniHard(tx, ty);
			return iss(ox, oy);
		}
		return false;
	}
	
	private static Coo upperLeftTest = new Coo();
	
	private boolean iss(int tx, int ty) {
		if (tiles.holdsPoint(tx, ty) && item == ROOMS().fData.item.get(tx, ty)) {
			ROOMS().fData.itemX1Y1(tx, ty, upperLeftTest, this);
			if (upperLeftTest.isSameAs(upperLeft))
				return true;
		}
		return false;
	}
	
	@Override
	public int mX(int tx, int ty) {
		iniHard(tx, ty);
		return dataCoo.x();
	}
	
	@Override
	public int mY(int tx, int ty) {
		iniHard(tx, ty);
		return dataCoo.y();
	}
	
	@Override
	public int x1(int tx, int ty) {
		iniHard(tx, ty);
		return tiles.x1();
	}
	
	@Override
	public int y1(int tx, int ty) {
		iniHard(tx, ty);
		return tiles.y1();
	}
	
	@Override
	public int width(int tx, int ty) {
		iniHard(tx, ty);
		return tiles.width();
	}
	
	@Override
	public int height(int tx, int ty) {
		iniHard(tx, ty);
		return tiles.height();
	}


}
